### PR TITLE
first draft of python neonUtilities instructions

### PR DESCRIPTION
**Pull request by cklunch**
@mjones01

### CODE REVIEW

File: tutorials/Python/neonUtilities/neonUtilitiesPython.md [hunk 1]
```diff
@@ -0,0 +1,342 @@
...
+1. Add C:\Program Files\R\R-3.3.1\bin\x64 to the Windows Environment Variable “Path”
+2. Install rpy2 manually from https://www.lfd.uci.edu/~gohlke/pythonlibs/#rpy2
+    1. Pick the correct version. At the download page the portion of the files with cp## relate to the Python version. e.g., rpy2 2.9.2 cp36 cp36m win_amd64.whl is the correct download when running Python 36 and 64 bit Windows (amd64).
+    2. Save the whl file, navigate to it in windows then run pip directly on the file as follows “pip install rpy2 2.9.2 cp36 cp36m win_amd64.whl”
+3. Add  an R_HOME Windows environment variable with the path C:\Program Files\R\R-3.4.3 (or whichever version you are running)
+4. Add an R_USER Windows environment variable with the path C:\Users\tgoulden\AppData\Local\Continuum\Anaconda3\Lib\site-packages\rpy2 (modify for your own user account)
+
+## Load packages
+
+Now import rpy2 into your session.
```
Comment (mjones01) [inline tutorials/Python/neonUtilities/neonUtilitiesPython.md]: Could add context for why we are pulling out the robjects and importr too.

File: tutorials/Python/neonUtilities/neonUtilitiesPython.md [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+
+Import the base R functionality.
+
+
+```python
+base = importr('base')
+utils = importr('utils')
+```
+
+Suppress R warnings. This step can be skipped, but will result in messages getting passed through from R that Python will interpret as warnings.
```
Comment (mjones01) [inline tutorials/Python/neonUtilities/neonUtilitiesPython.md]: What is the reason for having these instructions here instead up with with the other imports?

File: tutorials/Python/neonUtilities/neonUtilitiesPython.md [hunk 3]
```diff
@@ -0,0 +1,342 @@
...
+
+```python
+import warnings
+from rpy2.rinterface import RRuntimeWarning
+warnings.filterwarnings("ignore", category=RRuntimeWarning)
+```
+
+Install the `devtools` R package. The system will ask you to choose a CRAN mirror, select one close to your location. You only need to do this step once, although if you use this code regularly, re-installing periodically to get the newest version is a good idea.
+
+(The semicolon at the end of the line (here, and in some other function calls below) can be omitted. It suppresses a note indicating the output of the function is null. This note appears because these functions download or modify files on your local drive, but none of the data are read into the Python or R environments.)
```
Comment (mjones01) [inline tutorials/Python/neonUtilities/neonUtilitiesPython.md]: This is great info. I wouldn't put it in parentheses, instead just keep as text or if you think it is really seperate from the main point of the lesson, use the Data Tip styling.

File: tutorials/Python/neonUtilities/neonUtilitiesPython.md [hunk 4]
```diff
@@ -0,0 +1,342 @@
...
+estimatedTime: 
+packagesLibraries: rpy2
+topics: data-management,rep-sci
+languagesTool: python
+dataProduct: 
+code1: /Python/neonUtilities/neonUtilitiesPython.ipynb
+tutorialSeries: 
+urlTitle: utilities-python
+---
+
```
Comment (mjones01) [inline tutorials/Python/neonUtilities/neonUtilitiesPython.md]: Future, once we figure out where training data sets will be stored, let's include a small downloadable data set that can be used for this and the neonUtilities lesson.

File: tutorials/Python/neonUtilities/neonUtilitiesPython.md [hunk 5]
```diff
@@ -0,0 +1,342 @@
...
+Using the `devtools` package, install the `neonUtilities` package. As with `devtools`, you can theoretically skip this step in the future, but since we update `neonUtilities` frequently, we recommend re-installing regularly.
+Then import the package.
+
+
+```python
+devtools.install_github('NEONScience/NEON-utilities/neonUtilities');
+neonUtils = importr('neonUtilities')
+```
+
+## Stack data files
```
Comment (mjones01) [inline tutorials/Python/neonUtilities/neonUtilitiesPython.md]: For the neonUtilities lesson the headers were the function names (`stackBytable()`). It would be nice to be consistent but I could go either way: name (as in neonUtilities) or purpose (as here). Thoughts?

File: tutorials/Python/neonUtilities/neonUtilitiesPython.md [hunk 6]
```diff
@@ -0,0 +1,342 @@
...
+
+Before starting, you will need:
+1. Python 3 installed. It is probably possible to use this workflow in Python 2, but these instructions were developed and tested using 3.6.4.
+2. R installed. You don't need to have ever used it directly. We tested using R 3.4.3, but most other recent versions should also work.
+3. rpy2 installed. Run the line below from the command line. It won't run within Jupyter, the error message is shown for clarity. It also likely will not work on Windows, see next section if you are running Windows.
+
+
+```python
```
Comment (mjones01) [inline tutorials/Python/neonUtilities/neonUtilitiesPython.md]: How does code comments work in Python?  Should we add some in to show good coding practice?

File: tutorials/Python/neonUtilities/neonUtilitiesPython.md [hunk 7]
```diff
@@ -0,0 +1,342 @@
...
+
+## Stack data files
+
+The function `stackByTable()` in `neonUtilities` merges the monthly, site-level files the <a href="http://data.neonscience.org/home" target="_blank">NEON Data Portal</a> provides. Start by downloading the dataset you're interested in from the Portal. It will download as a single zip file. Note the file path it's saved to and proceed.
+
+The data stacker package comes with a data file, table_types. The data file is needed for the package to work, and rpy2 doesn't load data by default. So we need to load it to the session and then pass it back to the R environment.
+
+First, load the data file:
+
+
```
Comment (mjones01) [inline tutorials/Python/neonUtilities/neonUtilitiesPython.md]: Add a note about the symbol around `__rdata__` being two underscores together.  It would also be great for newer users to have some context about what each of the commands are doing.

File: tutorials/Python/neonUtilities/neonUtilitiesPython.md [hunk 8]
```diff
@@ -0,0 +1,342 @@
...
+Now, pass it back to the R environment:
+
+
+```python
+robjects.globalenv['table_types'] = table_types
+```
+
+Now run the `stackByTable()` function to stack the data. It requires two inputs: 
+1. The data product identifier (DPID) of the data you're stacking. DPIDs can be found in the <a href="http://data.neonscience.org/data-product-catalog" target="_blank">Data Product Catalog</a> and are in the form DP#.######.001
```
Comment (mjones01) [inline tutorials/Python/neonUtilities/neonUtilitiesPython.md]: I think we should be writing the data product number as DP#.######.### as it will be possible that .002 comes out at some point, correct.

File: tutorials/Python/neonUtilities/neonUtilitiesPython.md [hunk 9]
```diff
@@ -0,0 +1,342 @@
...
+2. A file path, the path to the zip file you downloaded from the NEON Data Portal.
+
+
+```python
+neonUtils.stackByTable('DP1.10100.001','/Downloads/NEON_isotope-soil-distrib-periodic.zip');
```
Comment (mjones01) [inline tutorials/Python/neonUtilities/neonUtilitiesPython.md]: I had to have a ~ in front of my file path.  Is that standard?  Or just variation by computer?

File: tutorials/Python/neonUtilities/neonUtilitiesPython.md [hunk 10]
```diff
@@ -0,0 +1,342 @@
...
+    
+
+
+Check the folder containing the original zip file from the Data Portal; you should now have a subfolder containing the unzipped and stacked files.
+
+## Download files to be stacked
+
+The function `zipsByProduct()` uses the <a href="http://data.neonscience.org/data-api" target="_blank">NEON API</a> to programmatically download data files for a given product. The files downloaded by `zipsByProduct()` can then be fed into `stackByTable()`.
+
+The function will create a new folder in the R working directory and write the files there. Set the working directory if it isn't set to where you want it.
```
Comment (mjones01) [inline tutorials/Python/neonUtilities/neonUtilitiesPython.md]: It isn't 100% clear to me if you are saying that the user needs to manually set the wd in R before the next step or if in the next step they will be setting the wd. I'm sure the later but clarification would be good.

File: tutorials/Python/neonUtilities/neonUtilitiesPython.md [hunk 11]
```diff
@@ -0,0 +1,342 @@
...
+
+
+```python
+base.setwd('/Downloads');
```
Comment (mjones01) [inline tutorials/Python/neonUtilities/neonUtilitiesPython.md]: Here too I needed a ~ in the file path.

File: tutorials/Python/neonUtilities/neonUtilitiesPython.md [hunk 12]
```diff
@@ -0,0 +1,342 @@
...
+```
+
+Run the downloader with these inputs: a DPID, a site (or "all" for all sites), a package (either basic or expanded), and an indicator to check the size of your download before proceeding, or not.
```
Comment (mjones01) [inline tutorials/Python/neonUtilities/neonUtilitiesPython.md]: Can you opt to download with the protocol documents?

File: tutorials/Python/neonUtilities/neonUtilitiesPython.md [hunk 13]
```diff
@@ -0,0 +1,342 @@
...
+
+There are two differences relative to running this function in R directly: 
+
+1. check.size becomes check_size, because dots have programmatic meaning in Python
+2. TRUE (or T) becomes "TRUE" because the values TRUE and FALSE don't have special meaning in Python the way they do in R, so it interprets them as variables if they're unquoted.
+
+check_size="TRUE" will estimate the size of the download and ask you to confirm before proceeding. This will cause problems in certain environments, or in batch processing. Under those circumstances, set check_size="FALSE", but consider the size of your query before doing this.
```
Comment (mjones01) [inline tutorials/Python/neonUtilities/neonUtilitiesPython.md]: The code bits here should be in `code font`

File: tutorials/Python/neonUtilities/neonUtilitiesPython.md [hunk 14]
```diff
@@ -0,0 +1,342 @@
...
+    
+
+
+The message output by `zipsByProduct()` indicates the file path where the files have been downloaded.
+
+Now take that file path and pass it to `stackByTable()`. The file structure is slightly different from the zip file returned by the Portal, so we need an additional input, folder="TRUE".
+
+
+```python
+neonUtils.stackByTable("/Downloads/filesToStack10023", folder="TRUE");
```
Comment (mjones01) [inline tutorials/Python/neonUtilities/neonUtilitiesPython.md]: Need to add in the DPID. As written, this command gives me this error: 
`Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/mjones01/anaconda3/lib/python3.6/site-packages/rpy2/robjects/functions.py", line 178, in __call__
    return super(SignatureTranslatedFunction, self).__call__(*args, **kwargs)
  File "/Users/mjones01/anaconda3/lib/python3.6/site-packages/rpy2/robjects/functions.py", line 106, in __call__
    res = super(Function, self).__call__(*new_args, **new_kwargs)
rpy2.rinterface.RRuntimeError: Error in (function (dpID, filepath, savepath = filepath, package = "basic",  : 
  ~/Downloads/filesToStack10023 is not a properly formatted data product ID. The correct format is DP#.#####.001, where the first placeholder must be between 1 and 4.`

File: tutorials/Python/neonUtilities/neonUtilitiesPython.md [hunk 15]
```diff
@@ -0,0 +1,342 @@
...
+    Copied the first available variable definition file to /stackedFiles and renamed as variables.csv
+    Copied the first available validation file to /stackedFiles and renamed as validation.csv
+    Stacked  hbp_massdata
+    Stacked  hbp_perbout
+    
+
+
+## Download remote sensing files
+
+The function `byFileAOP()` uses the <a href="http://data.neonscience.org/data-api" target="_blank">NEON API</a> to programmatically download data files for remote sensing (AOP) data products. These files cannot be stacked by `stackByTable()` because they are not tabular data. The function simply creates a folder in your working directory and writes the files there. It preserves the folder structure for the subproducts.
```
Comment (mjones01) [inline tutorials/Python/neonUtilities/neonUtilitiesPython.md]: General question: Will this function work for the bundles eddy covariance products in HDF5 format too?

File: tutorials/Python/neonUtilities/neonUtilitiesPython.md [hunk 16]
```diff
@@ -0,0 +1,342 @@
...
+2. R installed. You don't need to have ever used it directly. We tested using R 3.4.3, but most other recent versions should also work.
+3. rpy2 installed. Run the line below from the command line. It won't run within Jupyter, the error message is shown for clarity. It also likely will not work on Windows, see next section if you are running Windows.
+
+
+```python
+pip install rpy2
+```
+
+    
+    The following command must be run outside of the IPython shell:
```
[the remark this answers is not in the document]
Comment (cklunch) [reply tutorials/Python/neonUtilities/neonUtilitiesPython.md]: I can't, that's the error message reply from Python. I could edit it in the md, but it would get overwritten any time I regenerate the md.

File: tutorials/Python/neonUtilities/neonUtilitiesPython.md [hunk 17]
```diff
@@ -0,0 +1,342 @@
...
+    
+        $ pip install rpy2
+    
+    The Python package manager (pip) can only be used from outside of IPython.
+    Please reissue the `pip` command in a separate terminal or command prompt.
+    
+    See the Python documentation for more informations on how to install packages:
+    
+        https://docs.python.org/3/installing/
```
[the remark this answers is not in the document]
Comment (cklunch) [reply tutorials/Python/neonUtilities/neonUtilitiesPython.md]: This is also part of the error message reply from Python.

File: tutorials/Python/neonUtilities/neonUtilitiesPython.md [hunk 18]
```diff
@@ -0,0 +1,342 @@
...
+
+Before starting, you will need:
+1. Python 3 installed. It is probably possible to use this workflow in Python 2, but these instructions were developed and tested using 3.6.4.
+2. R installed. You don't need to have ever used it directly. We tested using R 3.4.3, but most other recent versions should also work.
+3. rpy2 installed. Run the line below from the command line. It won't run within Jupyter, the error message is shown for clarity. It also likely will not work on Windows, see next section if you are running Windows.
+
+
+```python
+pip install rpy2
```
[the remark this answers is not in the document]
Comment (cklunch) [reply tutorials/Python/neonUtilities/neonUtilitiesPython.md]: I think this is actually a pip error, added a line about needing pip installed. I could be wrong though.

File: tutorials/Python/neonUtilities/neonUtilitiesPython.md [hunk 19]
```diff
@@ -0,0 +1,342 @@
...
+languagesTool: python
+dataProduct: 
+code1: /Python/neonUtilities/neonUtilitiesPython.ipynb
+tutorialSeries: 
+urlTitle: utilities-python
+---
+
+## Install and set up
+
+The instructions below will guide you through using the neonUtilities R package in Python, via the rpy2 package. rpy2 creates an R environment you can interact with from Python. The focus in this tutorial is on the Python implementation, rather than a comprehensive overview of the package itself. For more information about the package, and instructions for running it in R directly, see the readme for the package on the <a href="https://github.com/NEONScience/NEON-utilities/tree/master/neonUtilities" target="_blank">NEON-utilities GitHub repo</a>, and the tutorial on the <a href="http://www.neonscience.org/neonDataStackR" target="_blank">NEON Data Skills page</a>.
```
[the remark this answers is not in the document]
Comment (cklunch) [reply tutorials/Python/neonUtilities/neonUtilitiesPython.md]: Do you know how to do that for numbered lists? I can adjust the text sections, but if I put a line return in a numbered list it all breaks.

File: tutorials/Python/neonUtilities/neonUtilitiesPython.md (hunk 2 above)
Comment (cklunch) [reply tutorials/Python/neonUtilities/neonUtilitiesPython.md]: I don't think I understand - this is in the setup section, along with the importing. ?

File: tutorials/Python/neonUtilities/neonUtilitiesPython.md (hunk 5 above)
Comment (cklunch) [reply tutorials/Python/neonUtilities/neonUtilitiesPython.md]: Good point. I like both. Possibly we could change both tutorials to something like "Stack data files with stackByTable()". Or is that too wordy?

File: tutorials/Python/neonUtilities/neonUtilitiesPython.md (hunk 8 above)
Comment (cklunch) [reply tutorials/Python/neonUtilities/neonUtilitiesPython.md]: The catalog doesn't include the REV number right now, so I think it will confuse people, they won't know what to put in the final 3 digits.

File: tutorials/Python/neonUtilities/neonUtilitiesPython.md (hunk 15 above)
Comment (cklunch) [reply tutorials/Python/neonUtilities/neonUtilitiesPython.md]: Not right now, that's functionality we could potentially add in the future. We haven't addressed the EC products in neonUtilities yet.

File: tutorials/Python/neonUtilities/neonUtilitiesPython.md (hunk 5 above)
Comment (mjones01) [reply tutorials/Python/neonUtilities/neonUtilitiesPython.md]: We could try.  "Combine Data Files: stackByTable()"?  Ideally we want to keep it to one line even on smaller desktop screens so the sidebar menu doesn't look too weird. See: http://www.neonscience.org/neon-api-usage

File: tutorials/Python/neonUtilities/neonUtilitiesPython.md (hunk 15 above)
Comment (mjones01) [reply tutorials/Python/neonUtilities/neonUtilitiesPython.md]: Okay.  Just wondering if it is designed for the non-csv products if it wouldn't be better to rename now, while there is other renaming going on and before the publicity push, to something without AOP in it.

File: tutorials/Python/neonUtilities/neonUtilitiesPython.md (hunk 16 above)
[the remark this answers is not in the document]
Comment (mjones01) [reply tutorials/Python/neonUtilities/neonUtilitiesPython.md]: Since this command shouldn't be run in Notebooks, can you suppress this code chunk from being evaluated (like eval=FALSE in RMD).  Then the error message wouldn't show and that better aligns with what the user will see as well.
